Handle delete action for VolumeOp

PatrickXYS opened the pull request:
**Which issue is resolved by this Pull Request:**
Resolves #1779 

**Description of your changes:**
Since VolumeOp doesn't support GC, we add support for VolumeOp delete operation, which means developers now can pass argument `action="delete"` to delete corresponding PVC. All necessary argument is given as an example below:

```python
dsl.VolumeOp(
        name="delete-pvc", # component's name
        resource_name="my-pvc", # The name of PVC need to delete
        action="delete" 
    )
```

Thus, after created a PVC for workflow, we can simply use delete VolumOp with ExitHandler to delete PVC.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/3194)
<!-- Reviewable:end -->

PatrickXYS commented:
/assign @Jeffwan

Jeffwan commented on sdk/python/kfp/dsl/_volume_op.py:
```diff
@@ -68,6 +68,37 @@ def __init__(self,
                             V1TypedLocalObjectReference)
         """
+        # condition when action == "delete"
+        if "action" in kwargs and kwargs["action"] == "delete":
```
I think it's fine to have `action` defined in `__init__`? Just use `create` as a default value and user can override it with `delete` value.

PatrickXYS replied:
It's a good point. I'll add it in the new revision.

Jeffwan commented on sdk/python/kfp/dsl/_volume_op.py:
```diff
@@ -68,6 +68,37 @@ def __init__(self,
+                k8s_resource=k8s_resource,
+                **kwargs,
+            )
+            return
```
This part is duplicate with existing code base. Is it possible to reuse them? I am thinking it should be fine to pass `delete` along to `ResourceOp` and let `ResourceOp` handle the real logic. We don't necessarily to have a branch here to handle this case.

PatrickXYS replied:
For this part, I'm thinking about using an elegant way to avoid duplication and increase reusability. Will propose a new revision to see how it goes.

PatrickXYS commented:
@Jeffwan I sent a new revision for addressing the comment. But I do have concerns about the latest commit, I think this commit somehow add more complexity with several `if action != "delete"` judgement. I'm not sure it's a better workaround. Could you check it out? 

Besides that, I'll squash all the commits after we think it's ok to test.

PatrickXYS commented:
I also find another issue: the PVC will not be deleted completed even though I triggered `action="delete"` in VolumeOp. 

Since PVC is protected with `finalizers: - kubernetes.io/pvc-protection`, to avoid it, we either patch the PVC with empty `finalizers` or manually delete all the pods related to the PVC. 

The details of this issue is described here: [k8s issue](https://github.com/kubernetes/kubernetes/issues/69697).

I think this might be another issue we need to solve after we finished this PR. WDYT? /cc @Jeffwan

Jeffwan commented on sdk/python/kfp/dsl/_volume_op.py:
```diff
@@ -44,6 +44,7 @@ def __init__(self,
                  modes: List[str] = None,
                  annotations: Dict[str, str] = None,
                  data_source=None,
+                 action: str = "create",
```
Can you add `action` to args documentation?

PatrickXYS replied:
Will resolve in the new revision.

Jeffwan commented:
> Since PVC is protected with finalizers: - kubernetes.io/pvc-protection, to avoid it, we either patch the PVC with empty finalizers or manually delete all the pods related to the PVC.

em. I am not sure even pods completes, the PVC is still considered being used by an `active` pod. If that's the problem, I don't think we still want to delete the PVC and leaving it in pending.

Jeffwan commented:
@Ark-kun @numerology Any suggestions?  The use case is we want to clean up volumes user created with VolumeOp in the pipeline, the challenge is if user don't want to  delete pod, then PVC can not be deleted because of the finializer. One workaround is to reset finalizer since pod is complete and we can then delete volume

PatrickXYS commented:
It makes sense about supporting a `ResourceOp.delete()`, for me, I think it makes more sense to having `VolumeOp.delete()`. Since people like using `VolumeOp` to create PVC, it's natural to use `VolumeOp` to delete the PVC. WDYT? @elikatsis @Jeffwan

PatrickXYS commented:
Btw, if the issue #1779 means delete PVC after we delete workflow pod, then `finalizers: - kubernetes.io/pvc-protection` is not an issue. Otherwise I need some suggestions from others. @hongye-sun

elikatsis commented:
Please take a look at #3213
